### PR TITLE
Sigmund

### PR DESCRIPTION
**Pull request by sigmunsl**
Added texture support

### CODE REVIEW

File: src/rasputin/geometry.py [hunk 1]
```diff
@@ -32,6 +32,10 @@ def __init__(self, *,
         self._aspects = None
         self._slopes = None
         self._colors = None
+        self._uvs = None
+        self.texture = np.empty((1024,1024))
+        self.scale_x = 1
+        self.scale_y = 1
```
Comment (skavhaug) [inline src/rasputin/geometry.py]: We should really do this part more properly

File: web/index.html [hunk 2]
```diff
@@ -22,6 +22,7 @@ <h1>Rasputin WebGL Visualizer</h1>
         <script src="js/three.js"></script>
         <script src="js/controls/OrbitControls.js"></script>
         <script src="data.js"></script>
-        <script src="index.js"></script>
+        <script type="module" src="https://threejsfundamentals.org/threejs/../3rdparty/dat.gui.module.js"></script>
+        <script type="module" src="index.js"></script>
```
Comment (skavhaug) [inline web/index.html:26]: Should we follow our prev pattern and include the resource?

File: web/index.js [hunk 3]
```diff
@@ -1,6 +1,12 @@
+import {GUI} from 'https://threejsfundamentals.org/threejs/../3rdparty/dat.gui.module.js';
```
Comment (skavhaug) [inline web/index.js:1]: Perhaps wait with the gui part to a different PR?

File: web/index.js [hunk 4]
```diff
@@ -15,18 +21,34 @@ function init({geometries}) {
     light2.position.set( 0, - 1, 0 );
     scene.add( light2 );
 
+    //const textureLoader = new THREE.TextureLoader();
+    //var texture = textureLoader.load("textures/sea.jpg");
+    //texture.wrapS = THREE.RepeatWrapping;
+    //texture.wrapT = THREE.RepeatWrapping;
+
```
Comment (skavhaug) [inline web/index.js]: Delete these lines

File: web/index.js [hunk 5]
```diff
@@ -15,18 +21,34 @@ function init({geometries}) {
...
+    //const gui = new GUI({width: 280});
+    //gui.add(new StringToNumberHelper(texture, 'wrapS'), 'value', wrapModes).name('Wrap Horizontal').onChange(updateTexture);
+    //gui.add(new StringToNumberHelper(texture, 'wrapT'), 'value', wrapModes).name('Wrap Vertical').onChange(updateTexture);
+    //gui.add(texture.repeat, 'x', 0, 5, .01).name('Repeat X');
+    //gui.add(texture.repeat, 'y', 0, 5, .01).name('Repeat Y');
+    //gui.add(texture.offset, 'x', -2, 2, .01).name('Offset X');
+    //gui.add(texture.offset, 'y', -2, 2, .01).name('Offset Y');
+    //gui.add(texture.center, 'x', -.5, 1.5, .01).name('Center X');
+    //gui.add(texture.center, 'y', -.5, 1.5, .01).name('Center Y');
+
```
Comment (skavhaug) [inline web/index.js:52]: Remove these lines

File: web/index.js [hunk 6]
```diff
@@ -47,41 +69,17 @@ function init({geometries}) {
...
-        uvs[3*i + 3] = (v2.y + offset.y)/range.y;
-        uvs[3*i + 4] = (v3.x + offset.x)/range.x;
-        uvs[3*i + 5] = (v3.y + offset.y)/range.y;
-    }
-    geometry.addAttribute("uv", new THREE.BufferAttribute(uvs, 2));
-    geometry.attributes.uv.needsUpdate = true;
-    return geometry;
+function create_material(geom, texture) {
+    var geomMaterial = new THREE.MeshPhysicalMaterial({map: texture, bumpMap: texture, vertexColors: THREE.FaceColors});
+    return new THREE.Mesh(geom, geomMaterial);
```
Comment (skavhaug) [inline web/index.js]: We don't need this function anymore?

**Review by skavhaug — CHANGES_REQUESTED** [review]
Some minor comments, plus one serious regarding scale and texture image size.

File: web/index.js (hunk 6 above)
Comment (sigmunsl) [reply web/index.js]: Fixed

File: web/index.js (hunk 5 above)
Comment (sigmunsl) [reply web/index.js:52]: Fixed

File: web/index.js (hunk 4 above)
Comment (sigmunsl) [reply web/index.js]: Fixed

**Review by skavhaug — APPROVED** [review]
LGTM!

**Comment by skavhaug** [comment]
We should really get some automatic testing in place for rasputin soon. We should make this our next effort!

**Pull request merged**